### DISCUSSION TITLE
From Extension into State question

### DISCUSSION THREAD

**Discussion by Paul-Vandell** [tokio-rs/axum#1558] [Q&A] [answered]
Hi there,
Since now `State` should be the recommended way instead of `Extension`
From the example [With the `extensions_mut`](https://docs.rs/axum/0.6.0-rc.5/axum/middleware/index.html#passing-state-from-middleware-to-handlers), The only way i succeed to get back my `current_user` was to **stick** like in the example with
`Extension(current_user): Extension<CurrentUser>`.

1. So i'm a bit confused for the future, will we have a kind of `states_mut` to alter it on fly instead of `extensions_mut` ?
2. Or does the `State` is just now the normal way of injecting a kind of  starting global `AppState` but `Extension` will still there knowing that it come from `tower` library ? 

Because now, we are agree that we can have this kind of function:
```
    async fn get_profile(
        Extension(current_user): Extension<CurrentUser>,
        State(state): State<AppState>,
    ) -> ApiResult<Profile> {
        Ok(Json(state.services.profile.from_user_id(&state.pool, current_user.id).await?))
    }
```

Maybe i have just misunderstood something here. It is just for clarification to lead me to the correct mindset and be prepare for the incoming features.

Thx for this awesome library.

**Comment by davidpdrsn** [comment, marked as answer]
State is for sharing data that is the same for all routes, like database connection pools. 

For storing request local data like authorization you still need to use extensions. Otherwise the user would be shared between all requests which isn't what you want.

**Reply by Paul-Vandell** [reply]
Got it, thanks :)
